### ISSUE TITLE
TypeError: Cannot read property 'native' of undefined

### ISSUE THREAD

**Issue by Eric8511** [log4js-node/log4js-node#968]
I use this module in a web project with node v10.16.0, npm 6.9.0, vue 2.9.6, and got an error when start the website:
index.js?8373:107 Uncaught TypeError: Cannot read property 'native' of undefined
    at eval (index.js?8373:107)
    at Object../node_modules/streamroller/node_modules/fs-extra/lib/fs/index.js (app.js:16953)
    at __webpack_require__ (app.js:767)
    at fn (app.js:130)
    at eval (index.js?215d:6)
    at Object../node_modules/streamroller/node_modules/fs-extra/lib/index.js (app.js:16965)
    at __webpack_require__ (app.js:767)
    at fn (app.js:130)
    at Object.eval (RollingFileWriteStream.js?9f14:2)
    at eval (RollingFileWriteStream.js:269)

**Comment by georgezouq** [comment]
same problem, any solution?

**Comment by nomiddlename** [comment]
Looks like you're trying to use webpack to bundle up code to use in a browser, but you're also configuring a rolling file stream - file streams won't work in the browser.

**Comment by bryanSME** [comment]
@nomiddlename is there no way to use this without the rolling file stream?

**Comment by nomiddlename** [comment]
I don't think that warning would come from using a file appender, because the file appenders are listed in the core appenders - which avoids the dependency-is-an-expression problem. Could you give me a bit more detail what version of log4js you're using, and what your config looks like, please?

**Comment by bryanSME** [comment]
Trying to make this as simple as possible, I've tried no config. Just adding the following causes the issue.  Running this in node has no issue for me. Running it on any code that hits the browser is where the warnings/errors start.  **Should I open a new ticket?**

version: 5.3.0 and 6.1.0

Here is the simplest code that causes the warning and failed compilation.
```
import { getLogger } from 'log4js';
getLogger(); 
```

WARNING in ./node_modules/log4js/lib/appenders/index.js 25:11-30
Critical dependency: the request of a dependency is an expression
 @ ./node_modules/log4js/lib/log4js.js
 @ ./src/...../........
 @ ./src/index.jsx
Child html-webpack-plugin for "index.html":
         Asset     Size  Chunks  Chunk Names
    index.html  537 KiB       1  
    Entrypoint undefined = index.html
    [./node_modules/html-webpack-plugin/lib/loader.js!./src/index.ejs] 2.69 KiB {1}
    [./node_modules/lodash/lodash.js] 528 KiB {1}
    [./node_modules/webpack/buildin/global.js] (webpack)/buildin/global.js 472 bytes {1}
    [./node_modules/webpack/buildin/module.js] (webpack)/buildin/module.js 497 bytes {1}
ℹ ｢wdm｣: Compiled with warnings.

**Comment by MkrishnaPodshore** [comment]
I am also facing the same issue and current log4js version I am using is "log4js": "^6.1.2", any solution for this issue.

Error Details:
index.js:107 Uncaught TypeError: Cannot read property 'native' of undefined
    at Object../node_modules/streamroller/node_modules/fs-extra/lib/fs/index.js (index.js:107)
    at __webpack_require__ (bootstrap:19)
    at Object../node_modules/streamroller/node_modules/fs-extra/lib/index.js (index.js:6)
    at __webpack_require__ (bootstrap:19)
    at Object.<anonymous> (RollingFileWriteStream.js:2)
    at Object../node_modules/streamroller/lib/RollingFileWriteStream.js (RollingFileWriteStream.js:267)
    at __webpack_require__ (bootstrap:19)
    at Object../node_modules/streamroller/lib/index.js (index.js:2)
    at __webpack_require__ (bootstrap:19)
    at Object.<anonymous> (file.js:3)

**Comment by mmaietta** [comment]
@MkrishnaPodshore, I've been hitting the same issue with using webpack, my final resort was a local patch. Since I'm targeting 'web' and the rolling file stream is not available in the first place, I used `patch-package` to simply ignore those appenders:

log4js+5.3.0.patch
```
diff --git a/node_modules/log4js/lib/appenders/index.js b/node_modules/log4js/lib/appenders/index.js
index ab78fcc..d418c2c 100644
--- a/node_modules/log4js/lib/appenders/index.js
+++ b/node_modules/log4js/lib/appenders/index.js
@@ -14,8 +14,8 @@ coreAppenders.set('stderr', require('./stderr'));
 coreAppenders.set('logLevelFilter', require('./logLevelFilter'));
 coreAppenders.set('categoryFilter', require('./categoryFilter'));
 coreAppenders.set('noLogFilter', require('./noLogFilter'));
-coreAppenders.set('file', require('./file'));
-coreAppenders.set('dateFile', require('./dateFile'));
+// coreAppenders.set('file', require('./file'));
+// coreAppenders.set('dateFile', require('./dateFile'));
 
 const appenders = new Map();
```

Presto, my console logging still works and fewer things are getting bundled.

**Comment by bryanSME** [comment]
Back on jan 30th, I contacted csausdev (module owner).  I think the communication was on slack but I can't find a log4j slack group.  All I have from the communication is this screenshot.  Hopefully this helps someone.  

![image](https://user-images.githubusercontent.com/16226280/79222898-5fd6ef80-7e26-11ea-9703-31101951b3f4.png)

**Comment by Download** [comment]
I made a very simple codepen that reproduces this issue:
https://codepen.io/StijnDeWitt/pen/rNxzOBE

There is a 'browserify cdn' that enables us to browserify any version of any module: https://wzrd.in
The codepen uses this to browserify log4js and load it onto the page:

```html
<script src="https://wzrd.in/standalone/log4js@6.3.0"></script>
```

We get this error:

```
Uncaught TypeError: Cannot read property 'native' of undefined
    at Object.89.graceful-fs (log4js@6.3.0:27)
    at i (log4js@6.3.0:1)
    at log4js@6.3.0:1
    at Object.90../copy (log4js@6.3.0:27)
    at i (log4js@6.3.0:1)
    at log4js@6.3.0:1
    at Object.<anonymous> (log4js@6.3.0:27)
    at Object.119../fileNameFormatter (log4js@6.3.0:27)
    at i (log4js@6.3.0:1)
    at log4js@6.3.0:1
```

I think we can solve this issue using the [`browser`](https://docs.npmjs.com/files/package.json#browser) field in package.json.
Note that the docs on NPM are incomplete. There is a better [specification of the `browser` field](https://github.com/defunctzombie/package-browser-field-spec) that basically boils down to there being 2 forms for it:

- String with the entry point for the browser version to replace `main` (as described in NPM docs)
- Object with file names of node versions of the code as keys and filenames of the browser versions as values

Browserify, Webpack, Rollup et all are aware of this `browser` field and it's specification and are using it to resolve `require` calls to the right version of the module. That's why you can use the same browserify CDN with `debug` [successfully](https://codepen.io/StijnDeWitt/pen/jOWLbrX). Browserify knows how to package `debug` because it [uses the `browser` field](https://github.com/visionmedia/debug/blob/4.1.1/package.json#L61) in package.json.
 
I think log4js could use the `browser` field to make 2 versions of the core appenders. One version would be used in Node and would have `file` appender in it and the other version would be used in the browsers and would not have the `file` appender.
I added a code review to #1023 suggesting this solution.

**Comment by ViieeS** [comment]
I'm looking for a good logger for client side and I'm fine with log4js on the server side.
Do I got right that it's incompatible with the client-side?

**Comment by Download** [comment]
For now, yeah. There is a [PR](https://github.com/log4js-node/log4js-node/pull/1023) in the works...

**Comment by lamweili** [comment]
Thanks to @aellerton who came up with PR #1374 which resolved the issue.
Will be released in `log4js@6.8.1`.

**Comment by lamweili** [comment]
Thanks to @aellerton who came up with PR #1374 which resolved the issue.
Will be released in `log4js@6.9.0`.
